### PR TITLE
Add addvariable function

### LINKED ISSUE #1172: constructconstraint! is not consistent with constructvariable!
The former creates a constraint that's *not* attached to the model, while the latter creates a variable that *is* attached to the model. Probably one should be renamed. This is also related to the inconsistency between Variable and ConstraintRef:
https://github.com/JuliaOpt/JuMP.jl/blob/2f6d6e570a37922f76606c99694a2b0d35b5629b/src/JuMP.jl#L268-L269

### PR DESCRIPTION
**Pull request by blegat**
Closes #1172

### CODE REVIEW

File: src/variables.jl [hunk 1]
```diff
@@ -3,6 +3,24 @@
...
+    upperbound::T
+    hasfix::Bool
+    fixedvalue::U
+    hasstart::Bool
+    start::V
+    binary::Bool
+    integer::Bool
+end
+
+struct ScalarVariable{S, T, U, V}
```
Comment (mlubin) [inline src/variables.jl]: `JuMP.Variable` needs to be renamed to `JuMP.VariableRef` before we can have an object called `ScalarVariable` with this meaning.

Comment (mlubin) [reply src/variables.jl]: (If you do that, make it a separate PR and then update this one afterwards.)

Comment (blegat) [reply src/variables.jl]: I have rebase the PR on master, it is now ready for review :)

File: src/macros.jl [hunk 2]
```diff
@@ -849,51 +849,15 @@ esc_nonconstant(x::Number) = x
...
-    fixedvalue::U
-    hasstart::Bool
-    start::V
-    binary::Bool
-    integer::Bool
-    name::String
-end
-
-# Returns the type of what `constructvariable!` would return with these starting positional arguments.
+# Returns the type of what `buildvariable` would return with these starting positional arguments.
```
Comment (mlubin) [inline src/macros.jl]: Does this still make sense? The positional arguments were replaced by the `VariableInfo` struct.

File: src/variables.jl [hunk 3]
```diff
@@ -3,6 +3,25 @@
...
```
Comment (mlubin) [inline src/variables.jl]: Is `ScalarVariable` supposed to inherit from `AbstractVariable`?

File: src/variables.jl [hunk 4]
```diff
@@ -389,3 +408,35 @@ resultvalue(v::VariableRef) = MOI.get(v.m, MOI.VariablePrimal(), v)
 hasresultvalues(m::Model) = MOI.canget(m, MOI.VariablePrimal(), VariableRef)
 
 @Base.deprecate setvalue(v::VariableRef, val::Number) setstart(v, val)
+
+"""
+    addvariable(m::Model, v::AbstractVariable, name::String="")
```
Comment (mlubin) [inline src/variables.jl]: The signature in the documentation should match the method you're defining. (`ScalarVariable` vs. `AbstractVariable`)

File: src/macros.jl [hunk 5]
```diff
@@ -849,51 +849,17 @@ esc_nonconstant(x::Number) = x
...
-    hasstart::Bool
-    start::V
-    binary::Bool
-    integer::Bool
-    name::String
-end
-
-# Returns the type of what `constructvariable!` would return with these starting positional arguments.
+# Returns the type of what `addvariable(::Model, buildvariable(...))` would return where `...` represents the positional arguments.
+# Example: `@variable m [1:3] foo` will allocate an vector of element type `variabletype(foo())`
```
Comment (mlubin) [inline src/macros.jl]: `variabletype(m, foo)`? (missing `m` and I think the parenthesis are incorrect)

**Pull request merged**